src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdsBidder.java

```diff
@@ -0,0 +1,1267 @@
+            if (openRTBRequest.getDevice().getConnectiontype() != null) {
+                network.type(openRTBRequest.getDevice().getConnectiontype());
+            } else {
+                network.type(HuaweiAdsConstants.DEFAULT_UNKNOWN_NETWORK_TYPE);
+            }
+
+            final List<CellInfo> cellInfos = new ArrayList<>();
+            if (StringUtils.isNotEmpty(openRTBRequest.getDevice().getMccmnc())) {
+                String[] arr = openRTBRequest.getDevice().getMccmnc().split("-");
+                network.carrier(0);
+                if (arr.length >= 2) {
+                    cellInfos.add(CellInfo.of(arr[0], arr[1]));
+                    String str = arr[0] + arr[1];
+                    switch (str) {
+                        case "46000", "46002", "46007" -> network.carrier(2);
+                        case "46001", "46006" -> network.carrier(1);
+                        case "46003", "46005", "46011" -> network.carrier(3);
+                        default -> network.carrier(99);
+                    }
+                }
+            }
+            network.cellInfo(cellInfos);
+            return network.build();
+        }
+        return null;
+    }
+
+    private Regs getRegsInfo(BidRequest openRTBRequest) {
+        if (openRTBRequest.getRegs() != null && openRTBRequest.getRegs().getCoppa() >= 0) {
+            return Regs.of(openRTBRequest.getRegs().getCoppa());
+        }
+        return null;
+    }
+
+    private Geo getGeoInfo(BidRequest openRTBRequest) {
+        if (openRTBRequest.getDevice() != null && openRTBRequest.getDevice().getGeo() != null) {
+
+            Float lon = openRTBRequest.getDevice().getGeo().getLon();
+            Float lat = openRTBRequest.getDevice().getGeo().getLat();
+
+            return Geo.of(lon != null ? BigDecimal.valueOf(lon) : null,
+                    lat != null ? BigDecimal.valueOf(lat) : null,
+                    openRTBRequest.getDevice().getGeo().getAccuracy(),
+                    openRTBRequest.getDevice().getGeo().getLastfix());
+        }
+
+        return null;
+    }
+
+    private String getConsentInfo(BidRequest openRtbRequest) {
+        if (openRtbRequest.getUser() != null && openRtbRequest.getUser().getExt() != null) {
+            final ExtUser extUser = openRtbRequest.getUser().getExt();
+
+            return extUser.getConsent();
+        }
+        return null;
+    }
+
+    private MultiMap getHeaders(ExtImpHuaweiAds huaweiAdsImpExt, BidRequest request) {
+        final MultiMap headers = HttpUtil.headers();
+
+        if (huaweiAdsImpExt == null) {
+            return headers;
+        }
+
+        final String authorization = getDigestAuthorization(huaweiAdsImpExt);
+
+        headers.set(HttpUtil.AUTHORIZATION_HEADER, authorization);
+
+        if (request.getDevice() != null
+                && request.getDevice().getUa() != null
+                && !request.getDevice().getUa().isEmpty()) {
```

Comment:
Call chains like this should be replaced with either chains of local variables(as we did prior to optionals) or with `Optional`